server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java

```diff
@@ -26,11 +26,92 @@
- * manage state/timeouts while the handshake is in transit, and handle the eventual response.
+ * Sends and receives transport-level connection handshakes. This class will send the initial handshake, manage state/timeouts while the
+ * handshake is in transit, and handle the eventual response.
  */
 final class TransportHandshaker {
 
+    /*
+     * The transport-level handshake allows the node that opened the connection to determine the newest protocol version with which it can
+     * communicate with the remote node. Each node sends its maximum acceptable protocol version to the other, but the responding node
+     * ignores the body of the request. After the handshake, the OutboundHandler uses the min(local,remote) protocol version for all later
+     * messages.
+     *
+     * This version supports two handshake protocols, v6080099 and v7170099, which respectively have the same message structure as the
+     * transport protocols of v6.8.0 and v7.17.0. This node only sends v7170099 requests, but it can send a valid response to any v6080099
+     * requests that it receives.
+     *
+     * Here are some example messages, broken down to show their structure:
+     *
+     * ## v6080099 Request:
+     *
+     * 45 53                            -- 'ES' marker
+     * 00 00 00 34                      -- total message length
+     *    00 00 00 00 00 00 00 01       -- request ID
+     *    08                            -- status flags (0b1000 == handshake request)
+     *    00 5c c6 63                   -- handshake protocol version (0x5cc663 == 6080099)
+     *    00                            -- no request headers [1]
+     *    00                            -- no response headers [1]
+     *    01                            -- one feature [2]
+     *       06                         -- feature name length
+     *          78 2d 70 61 63 6b       -- feature name 'x-pack'
+     *    16                            -- action string size
+     *       69 6e 74 65 72 6e 61 6c    }
+     *       3a 74 63 70 2f 68 61 6e    }- ASCII representation of HANDSHAKE_ACTION_NAME
+     *       64 73 68 61 6b 65          }
+     *    00                            -- no parent task ID [3]
+     *    04                            -- payload length
+     *       8b d5 b5 03                -- max acceptable protocol version (vInt: 00000011 10110101 11010101 10001011 == 7170699)
+     *
+     * ## v6080099 Response:
+     *
+     * 45 53                            -- 'ES' marker
+     * 00 00 00 13                      -- total message length
+     *    00 00 00 00 00 00 00 01       -- request ID (copied from request)
+     *    09                            -- status flags (0b1001 == handshake response)
+     *    00 5c c6 63                   -- handshake protocol version (0x5cc663 == 6080099, copied from request)
+     *    00                            -- no request headers [1]
```

Comment:
It looks like this might be a full int? https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/transport/InboundDecoder.java#L188

Reply:
This is a v6080099 response so `remoteVersion.onOrAfter(TcpHeader.VERSION_WITH_HEADER_SIZE)` is false here. You're thinking of this line instead:

https://github.com/elastic/elasticsearch/blob/f786997aab9ee0db0b608f315ebc56347a894ffe/server/src/main/java/org/elasticsearch/transport/TransportHandshaker.java#L104

Reply:
:facepalm: